backupccl: fix error when the restore rollback cleanup validates uncommitted schema descriptor

gh-casper opened the pull request:
Previously, during a restore rollback, we only marked database as dropped to skip pre-txn validation when we drop all descriptors that we created at the start of the restore. This is not sufficient because we will also run validation against user-defined schemas.

This patch marked schema as dropped to skip validation too.

Fixes: #70049
Release note: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/70551)
<!-- Reviewable:end -->

gh-casper committed:
backupccl: fix error when the restore rollback cleanup validates uncommitted schema descriptor

Previously, during a restore rollback, we only marked database as dropped to skip pre-txn validation when we drop all descriptors that we created at the start of the restore. This is not sufficient because we will also run validation against user-defined schemas.

This patch marked schema as dropped to skip validation too.

Fixes: cockroachdb#70049
Release note: none.

adityamaru commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -2234,24 +2234,36 @@ func (r *restoreResumer) dropDescriptors(
+		b.Del(catalogkeys.EncodeNameKey(codec, mutSchema))
+		b.Del(catalogkeys.MakeDescMetadataKey(codec, mutSchema.GetID()))
+		descsCol.AddDeletedDescriptor(mutSchema)
+		dbsWithDeletedSchemas[mutSchema.GetParentID()] = append(dbsWithDeletedSchemas[mutSchema.GetParentID()], mutSchema)
```
I think we should store to an immutable schema descriptor in this map. Even though it doesn't change much it prevents us from unintentionally mutating the descriptor wherever we use the values in this map.

gh-casper replied:
No worry in this case because what we are adding into `dbsWithDeletedSchemas` map is catalog.Descriptor which only has immutable APIs.

adityamaru reviewed (approved):
Nice!

gh-casper commented:
bors r+
